### PR TITLE
Use StringView for protocol stack to avoid calling strlen on strings with a known length.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This is an alternative proposal to #1520.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
I checked the compiler output and AFAICT passing a `StringView` by value is the same cost as passing two integers. That is, passing a `StringView` by value has the same performance as passing a pointer and a length by value.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/freebsd-github/1665/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/97/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1561/ for details.

**Comment by atsci** [comment]
FreeBSD build *failed*! See https://ci.trafficserver.apache.org/job/freebsd-github/1666/ for details.

**Comment by atsci** [comment]
clang-analyzer build *failed*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/229/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1562/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/98/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1563/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/99/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/freebsd-github/1667/ for details.

**Comment by atsci** [comment]
clang-analyzer build *failed*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/230/ for details.

**Comment by atsci** [comment]
clang-analyzer build *failed*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/231/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/100/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/freebsd-github/1668/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1564/ for details.

**Comment by atsci** [comment]
clang-analyzer build *successful*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/232/ for details.

**Comment by shukitchan** [comment]
👍 
Looks good. 
Just one question. Should the internal "protocol_contains" functions return "StringView" as well? 
Obviously we should not be changing the return type of public API. But internally should we be using "StringView" for those internal api calls?

**Comment by SolidWallOfCode** [comment]
Hmmmm. I think I'd leave that for another day. Those aren't really used internally, only to forward the API request. The return value isn't examined except to check for `nullptr` which is just as fast without the `StringView`. Further the `StringView` can't be returned via the TS API without changes, making that not particularly useful. A key goal of that call is to normalize the string to allow direct pointer comparison, in which case the length is superflous.

**Comment by shukitchan** [comment]
👍

**Comment by zwoop** [comment]
It even works:

```
odin (08:54) 283/0 $ curl -D - -o /dev/null -s  https://www.ogre.com/ | grep Via:
Via: http/1.1 tls/1.2 tcp ipv4 Cosmo (ApacheTrafficServer/7.2.0 [cRs f ])
```

Question: Do we really want to modify the Via: header format in the mid-release cycle? I guess i'm ok with it, but, a concern would be if anyone parses the Via: header for whatever reason, it might break things?

File: proxy/http/HttpSM.cc [hunk 1]
```diff
@@ -8057,14 +8059,13 @@ HttpSM::is_redirect_required()
 
 // Fill in the client protocols used.  Return the number of entries returned
 int
-HttpSM::populate_client_protocol(const char **result, int n) const
+HttpSM::populate_client_protocol(ts::StringView *result, int n) const
```
Comment (maskit) [inline proxy/http/HttpSM.cc]: StringView (w/o namespace) ?

**Comment by mlibbey** [comment]
I'm terrible at reading RFC docs, so, fwiw, the Via header structure is described here:
https://tools.ietf.org/html/rfc7230#section-5.7.1

File: proxy/http/HttpTransactHeaders.cc [hunk 2]
```diff
@@ -685,6 +687,23 @@ HttpTransactHeaders::insert_server_header_in_response(const char *server_tag, in
   }
 }
 
+/// Look up the protocol stack and write it to the @a via_string.
+size_t
+Write_Client_Protocol_Stack(HttpTransact::State *s, char *via_string, size_t len)
```
Comment (maskit) [inline proxy/http/HttpTransactHeaders.cc]: write_client_protocol_stack? Is this intentional?

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/RAT-github/1/ for details.

**Comment by SolidWallOfCode** [comment]
Upon further review of this, #1520, and #1506, I think this is just a bug fix, in that there was already code in place to report the protocol stack in the VIA header. Therefore this is primarily an update to use `StringView` (and incidentally fix the #1506 bug).

**Comment by atsci** [comment]
FreeBSD11 build *failed*! See https://ci.trafficserver.apache.org/job/freebsd-github/1700/ for details.

**Comment by atsci** [comment]
clang-format check *successful*! See https://ci.trafficserver.apache.org/job/clang-format-github/5/ for details.

**Comment by atsci** [comment]
RAT check *successful*! See https://ci.trafficserver.apache.org/job/RAT-github/18/ for details.

**Comment by atsci** [comment]
AU check *successful*! See https://ci.trafficserver.apache.org/job/autest-github/4/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1595/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/131/ for details.

**Comment by atsci** [comment]
clang-analyzer build *successful*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/263/ for details.

**Comment by zwoop** [comment]
Gonna build it again anyways ... [approve-ci]

**Comment by atsci** [comment]
clang-format check *successful*! See https://ci.trafficserver.apache.org/job/clang-format-github/6/ for details.

**Comment by atsci** [comment]
RAT check *successful*! See https://ci.trafficserver.apache.org/job/RAT-github/19/ for details.

**Comment by atsci** [comment]
FreeBSD11 build *failed*! See https://ci.trafficserver.apache.org/job/freebsd-github/1701/ for details.

**Comment by atsci** [comment]
AU check *failed*! See https://ci.trafficserver.apache.org/job/autest-github/5/ for details.

**Comment by atsci** [comment]
Intel CC build *successful*! See https://ci.trafficserver.apache.org/job/icc-github/132/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/linux-github/1596/ for details.

**Comment by atsci** [comment]
clang-analyzer build *successful*! See https://ci.trafficserver.apache.org/job/clang-analyzer-github/264/ for details.

**Comment by maskit** [comment]
It looks Http2ClientSession::PROTO_TAG is empty to me. I see two spaces between "http/1.1" and "tls/1.2".

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/16/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/29/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/15/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/142/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1606/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1711/

**Comment by SolidWallOfCode** [comment]
@maskit I'll look at that, but zwoop's test output has the 'h2' in it.

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/274/

File: proxy/http/HttpSM.cc (hunk 1 above)
Comment (SolidWallOfCode) [inline proxy/http/HttpSM.cc]: This should be fixed.

File: iocore/net/P_UnixNetVConnection.h [hunk 3]
```diff
@@ -149,21 +149,21 @@ class UnixNetVConnection : public NetVConnection
...
-  virtual bool add_to_active_queue();
+  void set_active_timeout(ink_hrtime timeout_in) override;
+  void set_inactivity_timeout(ink_hrtime timeout_in) override;
+  void cancel_active_timeout() override;
+  void cancel_inactivity_timeout() override;
+  void set_action(Continuation *c) override;
+  void add_to_keep_alive_queue() override;
+  void remove_from_keep_alive_queue() override;
+  bool add_to_active_queue() override;
   virtual void remove_from_active_queue();
```
Comment (SolidWallOfCode) [inline iocore/net/P_UnixNetVConnection.h]: @bryancall Note `remove_from_active_queue` does not override anything, in contrast to the rest of this group of methods.

**Comment by maskit** [comment]
I can see "h2" if I build it on Linux with gcc but not on Mac with clang.

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/22/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/35/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/21/

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1612/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/22/

**Comment by atsci** [comment]
Linux build *failed*! https://ci.trafficserver.apache.org/job/linux-github/1613/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/23/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/36/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1717/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/147/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/148/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/25/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/38/

**Comment by atsci** [comment]
Intel CC build *failed*! https://ci.trafficserver.apache.org/job/icc-github/150/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/23/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/26/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/39/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1718/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1614/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/25/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/279/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/151/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1720/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1721/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/280/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1616/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/282/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/283/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/31/

**Comment by atsci** [comment]
RAT check *successful*! https://ci.trafficserver.apache.org/job/RAT-github/44/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/30/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1726/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1621/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/156/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/288/

**Comment by SolidWallOfCode** [comment]
@bryancall says `StringView` should be changed to be conformant with `string_ref` from C++17. This would involve changing the method naming scheme and adding some additional methods to cover those in `string_ref`.

**Comment by SolidWallOfCode** [comment]
Squashing and verifying.

**Comment by atsci** [comment]
RAT check *failed*! https://ci.trafficserver.apache.org/job/RAT-github/53/

**Comment by atsci** [comment]
clang format *successful*! https://ci.trafficserver.apache.org/job/clang-format-github/40/

**Comment by atsci** [comment]
AU check *failed*! https://ci.trafficserver.apache.org/job/autest-github/38/

**Comment by atsci** [comment]
FreeBSD11 build *successful*! https://ci.trafficserver.apache.org/job/freebsd-github/1735/

**Comment by atsci** [comment]
Intel CC build *successful*! https://ci.trafficserver.apache.org/job/icc-github/164/

**Comment by atsci** [comment]
Linux build *successful*! https://ci.trafficserver.apache.org/job/linux-github/1630/

**Comment by atsci** [comment]
clang-analyzer build *successful*! https://ci.trafficserver.apache.org/job/clang-analyzer-github/296/

**Pull request merged**